### ISSUE TITLE
local-global repositories (composer link command)

### ISSUE THREAD

**Issue by fprochazka** [composer/composer#601]
Hi,
 I'm writting simulatenously few libraries and it's realy pain to commit everything and for every few lines write <code>$ composer update</code>. So it would be nice to have new kind of local repository or something, that wouldn't copy the dependency into project, but rather link to it. So when I change something in one project, it affects the other one immediately and I don't have to commit stuff and run update on composer.

Some command like "link this library, but don't touch it, just autoload it".

Or is there any suggested worflow on this problem?

**Comment by dantleech** [comment]
This would be a nice feature..

**Comment by minutephp** [comment]
Yes, this is a long sought after feature.. right now I'm doing this to get a similar result (see the "Brave new Local Development" part)

http://gos.si/blog/composer-development-with-local-dependencies

**Comment by rrajkomar** [comment]
@minutephp : What you suggest in your post (which I had already read prior to creating issue #4152) is nice but would only ork for the current version you have on your machine (if you switch branches dependecies  content get switched as well).
Using the repositories option in the ~/.composer/config.json file, you can achieve the same result, you would also be able to handle branches and tags without issues (when using git, AFAIK  it would not work with other VCS, i.e. with svn you can't do it because there's no local repository)  but as mentionned in the issue it is not very intuitive.

Maybe a solution would be to have another configuration option that would work the other way around : the global config could override values defined in the project's composer.json

For the moment, since I'm working mainly with git, I can work around the issue with the reposiotires option set only in each developer's config.json (and using the same method on the different servers I use).

**Comment by ahuszko** [comment]
This may help you to improve local package development: http://ahuszko.ghost.io/compact/

**Comment by staabm** [comment]
@Seldaek I think we can close this one, because we have `PathRepository` now?
https://getcomposer.org/doc/05-repositories.md#path

**Comment by Seldaek** [comment]
No I'd still like to investigate a better solution.. path repo has several issues IMO it's a good simple solution but not really the best.

**Comment by Petah** [comment]
+1

**Comment by Seldaek** [comment]
@Petah there is a subscribe button on the right to avoid spamming people with useless +1s.

**Comment by Petah** [comment]
Thanks :+1:

**Comment by mreschke** [comment]
The `path repo` can get tripped up sometimes for odd reasons with versioning...not a bug, just logic I can't figure out.  But for me the `path repo` works amazingly for package development.  @Seldaek what are some of the issues you are seeing?

**Comment by rudiedirkx** [comment]
As I see it (I'm new) _path repo_ only works for the owner/primary developer, since the location is in the shared `composer.json`. This should be a dev config IMO, which means another (optional) file that is not shared, like `composer-dev.json`.

I don't understand at all why the package is symlinked or copied. Why not just read from the source?

#1299 is full of good ideas IMO:

* separate file
* no symlinks
* `$autoloader->setPsr4()` as a back-up, which helps me now

Sorry if I'm repeating discarded/irrelevant ideas.

**Comment by rrajkomar** [comment]
@mreschke : Actually the path repository type does not match the requirement at all since the idea is not to add another source for the dependencies but simply to be able to tell that on dev machines private packages dependencies should be retrieved locally instead of through the otherwise used composer typed repository. Therefore the project composer.json file structure should not have to suffer changes for this feature support.

@rudiedirkx : Using another file that would not be commited (much in the same way as symfony's parameters.yml file) could be a solution if Composer could prioritize it as the first place it has to look into for packages before the usual (/comitted) composer.json file. Although there would be the composer.lock (if present) mismatch issue to handle as well.

**Comment by Seldaek** [comment]
Another util by @franzliedke that seems to be a good fit for this use case: https://github.com/franzliedke/studio

**Comment by franzliedke** [comment]
Thanks for mentioning me, @Seldaek. This is, in fact, precisely the use case that Studio makes possible.

With the latest version of Composer, I can finally prepare the new release and will post a proper announcement on my blog.

**Comment by Basster** [comment]
Thanks, @Seldaek and @franzliedke. Exactly was I was looking for in #4125. I'm happy now. :)

**Comment by b01** [comment]
@Seldaek I'd been having problems with using ~/.composer/config.json "path" method on my Macbook Pro for some strange reason. Using @francisbesset Studio worked with no issue.

**Comment by francisbesset** [comment]
@b01 not me but @franzliedke ;)

**Comment by fboutin-pmc** [comment]
I was looking for a `composer link` command as well. `Studio` looked good but I think it's pretty bloated for what it does. So we finally ended-up with manually make use of [Symbolic Links](https://en.wikipedia.org/wiki/Symbolic_link) combined with this handy script that re-deploy linked components,

```
composer run-script post-install-cmd -vvv -- --redeploy
```

Put it simply,

1. You obviously `git clone` the COMPONENT repo you want to link
1. In the project where you make use of the component, you `rm -R vendor/COMPONENT`
1. Followed with, `ln -s FULL_PATH_OF_COMPONENT_REPO vendor/COMPONENT
1. You run the aforementioned script and voilà!

**Comment by rudiedirkx** [comment]
@fboutin-pmc That's what I do too, but with a [global package](https://packagist.org/packages/rdx/composer-localoader-plugin) that maintains the links per project.

**Comment by franzliedke** [comment]
@fboutin-pmc Care to explain why you think Studio is bloated? I would obviously love to know how, and if I can change that. I'd say feature-completeness is a bigger problem. ;)

**Comment by atrauzzi** [comment]
Up front, the requirement of having to define anything in a consuming repository to support link functionality is a little inconvenient.  I think that's the beauty of `npm link` really -- it's part of the ecosystem.

**Comment by nerdbeere** [comment]
> the requirement of having to define anything in a consuming repository to support link functionality is a little inconvenient

I can't stress this enough. I have to admit I'm pretty new to the whole PHP / Composer Ecosystem, but coming from `npm`, developing packages locally without a proper `link` command is a real pain for me.

The beauty about `npm link` is, that it doesn't require any changes to your code. Maybe that's just me, but when I work on a feature I only want to see changes in files that actually belong to this feature.

**Comment by atrauzzi** [comment]
@iamandrewluca - While this symptomatically gets the desired outcome, we need a way to do this without having to alter the `composer.json` file and risk accidentally committing these overrides.

This is why the major comparison here is to `npm link`.

**Comment by dave-newson** [comment]
As it's been about 10 months and this thread has no conclusion, so I'd like to give a bit of an overview, especially for anyone else coming to visit this thread, as I did, looking for an equivalent to `npm link`

### TL;DR

I believe the current best-of-class option is provided by [Studio](https://github.com/franzliedke/studio), which adds to your PHP toolchain beside composer and can provide an `npm link` esq behavior.

## Overview

### The official Composer mechanism

https://johannespichler.com/developing-composer-packages-locally/
https://gos.si/blog/composer-development-with-local-dependencies/

As pointed out by @iamandrewluca - the official mechanism involves modifying the project's composer.json to establish a local link, including specifying the path there the library can be found.

People feel this approach is flawed because:
- It requires each project to opt-in to this linked version of the library
- It requires an edit of a file tracked by VCS

### Do it manually

You could also jump into the dirs and set up the symlinks yourself in the vendors dir, but that's a whole bunch of effort and I _believe_ we're programmers and against the idea of making humans do the work of machines.

### composer-localdev-plugin

https://github.com/gossi/composer-localdev-plugin

This one is like the infant-version of what most of us are after - similar behavior to `npm link`, only it's defined in a configuration file in your home directory.

Issues include:
 - It's an external plugin people need to discover exists (boo)
 - Setting up the config is a little fiddly.

### composer-localoader-plugin

https://github.com/rudiedirkx/composer-localoader-plugin

This is almost exactly like the above one, only it includes a command, equivalent to `npm link`

### studio

https://github.com/franzliedke/studio

This is kind of a swiss-army-knife of package development, similar but more powerful than composer-localdev-plugin by making it a good way to create/link/unlink/destroy your packages.

The primary issue with studio is I had no idea it existed until I read this thread.  I guess the only solution here is to talk about it more (hello world) and maybe write a blog post singing its praises in the misguided assumption that people listen to me.

We still have to install this is a global lib, so in my use case it'll become part of our toolchain beside composer, and I'll be making all my devs install it if they want to do package development.

## My Use Case

Just to clarify why i'm here, and add to the voices that think composer should include this kind of functionality out of the box:

I'm writing a library that's framework-agnostic (ala Doctrine).

To plug this library into various frameworks (Symfony2, Symfony3, Symfony4, Symfony Flex, Laravel) I need to write various Bundles/libs to wrap my framework-agnostic lib, and I also need to set up mini-applications so I can test these wrappers against 5+ different frameworks.

In all, this results in ~6 applications on my local machine which I want to use my local version of code, rather than pull the lib every time.

With a feature like `npm link` this is attained in a matter of minutes - simply link the dependency, running composer install for my various test apps (all automated, thanks) and I can continue with my day-job.

With the current composer-suggested method I have to change 6+ composer.json files and remember not to commit them - as does everyone on my team. If this method were suggested by another dev in my department I would laugh and reply with "ew no there must be a better way".

As it happens, Studio provides that method, however I had to crawl through this thread and examine all the options to find it.

I hope this post will serve to save people the 40 minutes I've just spent on this, and maybe as a nudge to the Composer team to investigate this feature further, nearly 7 years after it was originally proposed.

**Comment by Seldaek** [comment]
@dave-newson as far as I can tell, your use case would be very well served by loading your local version of your lib through a [path repository](https://getcomposer.org/doc/05-repositories.md#path) as a symlink. I see that studio and co can do more, and maybe Composer will one day too, but for that particular use case path repos are more than enough.

**Comment by franzliedke** [comment]
@Seldaek I think his point was that path repositories require modifications to versioned files, which can be undesirable when using this setup only locally to e.g. test new versions of the linked library. Studio does quite a bit of work to avoid that.

@dave-newson Thanks for the kind words. A blog post has been on my to-do list for far too long. :see_no_evil:

**Comment by Seldaek** [comment]
@franzliedke I understand that's the main benefit yes, but it sounds to me like he has a few local test apps that wouldn't really suffer from that problem.

**Comment by dave-newson** [comment]
@Seldaek For me it's really a question of scale. 

Yes - you're not mistaken - I can go into each of these libraries and I can modify VCS-tracked files (ew) to point to specific paths (ew) on my local machine, and it _will_ work.

However, the more libraries (2) and apps (4) that I have to do this with increases the likelihood of mistakes, human error, and wasted time.  If you then scale this up by the number of developers I have in the office who also need to work on this code (3) you get.. 2x4x3 = 24 _manual edits_

That.. well that's _silly_ . We're not ever doing that, and we'll find an alternative route (composer-localdev-plugin) if Composer won't support our use case.

And while I've fixed this problem by going to a third party, this question of scale expands to all libraries in the wild who want the same kind of behavior - that, to me, looks like a serious black-hole of man-hours.

Rather than continue to argue in a circle, can I ask:

 - Are you still open to supporting a global library-symlinking feature, similar to that of composer-localdev-plugin with a command like `composer link` ([see comment from 2012](https://github.com/composer/composer/issues/601#issuecomment-5229994))?
 - If yes - as I'm not overly familiar with composer internals - can you think of any pitfalls to implementing this?

**Comment by hacktivista** [comment]
FYI https://getcomposer.org/doc/03-cli.md#composer-home-config-json

**Comment by atrauzzi** [comment]
@dave-newson - 💯 The truth is that based on what's needed here, the workarounds and alternatives suggested don't meet the need and are not an undertaking I'd share as a best practice with my team.

I recently aliased a package using yarn -- it was such a cinch and didn't get in my way at all.  There was virtually no setup, just one command in the source repo and another to bind it in the consuming project.

The truth is, this is a feature that would serve the community far more than is being initially gauged by reducing friction when working with packages in-context.  Think of how much easier it will be to tell people how to boot a local checkout of a project to submit bug fixes **from within the projects they're experiencing the issues in**!  Heck, it might even help prevent bugs by encouraging people to understand the libraries they consume better.

Those two reasons alone are enough to make this worth having as an official feature in composer.

**Comment by esynaps** [comment]
The solution proposed by @hacktivista is not sufficient.

> This file allows you to set repositories and configuration for the user's projects.
> 
> In case global configuration matches local configuration, the local configuration in the project's composer.json always wins.

You can not overwrite a VCS repository (used by the remote) and défined in composer.json with a path repository defined in config.json

**Comment by hacktivista** [comment]
@esynaps you can set a path as VCS

**Comment by hacktivista** [comment]
@esynaps... you could define a global VCS repo in your local and a global VCS repo in your remote... as in https://getcomposer.org/doc/05-repositories.md#using-private-repositories

**Comment by atrauzzi** [comment]
@hacktivista - That still doesn't solve the fact that it requires a lot of extra machine-specific setup.

This approach isn't a _solution_ to the original problem.  Symptomatically obtaining an outcome is very different than ascertaining the spirit of a desired feature and not creating extra work or scalability problems along the way.

**Comment by hacktivista** [comment]
Still it's a valid solution for some use-cases.

**Comment by atrauzzi** [comment]
I'll put here what I usually say in response to _"workaround"_ answers: Be careful you aren't hijacking the discussion or giving maintainers a bad signal (that the issue is not valid).

It would be more on topic to not reiterate the same workaround more than once.  In fact, that's why I'm replying still -- to assert that what's been brought up so far (including workarounds) doesn't resolve this ticket.  That way nothing gets misinterpreted on a skim (it happens 😢).

**Comment by gnutix** [comment]
This issue is still a huge PITA, and neither https://github.com/gossi/composer-localdev-plugin nor https://github.com/franzliedke/studio completely work for me (I spent two days trying to get a proper setup with a working autoloading, unspoiled composer.json/lock files, and no mess with composer branches in the vendors... no success). There's tons of blog posts everywhere since years providing pseudo workarounds, but nothing stands.

**Comment by lkraav** [comment]
I landed on this issue when my "just work on vendor clone" dependency package defined a new dependency.

Running `composer update` to install the new sub-dependency in main app can't go forward, because it looks for `dev-master` updates on GitHub, which I don't want to push vendor clone wip code to.

Fortunately `composer` detected unpushed changes and at least didn't unexpectedly destroy work :+1: 

Like others before, I'm also thinking some kind of a "silent local link" mechanism to avoid changing version-controlled `composer.{json,lock}` files would be useful.

Dec 2019: Studio still looks like the most advanced package available to take care of this problem.

**Comment by lkraav** [comment]
https://github.com/Letudiant/composer-shared-package-plugin doesn't seem to be presented in this issue before.

**Comment by pawawat** [comment]
any update on this problem?

**Comment by henzeb** [comment]
In case anyone lands here, None of the above mentioned packages worked for me the way I wanted it, so I wrote my own. Should work for composer 1 and 2. Currently at Alpa, but I've been testing it for a week, but seems stable. 

It goes back to the very basics, just replace a package with a symlink (it means the package should already exist and required in your project), without touching composer.json or lock files.

https://packagist.org/packages/henzeb/composer-link

**Comment by gmazzap** [comment]
Hi @henzeb thanks for your work.

There's the discussion about to bring "link" functionality natively in Composer.

The discussion started 8+ years ago here: #601 and recently I'm trying to move it forward. I've written a requirements document on how it could work, but I wrote no code yet waiting for maintainers feedback.

Considering the experience you gained on the topic, if you could find the time to give some feedback on https://gist.github.com/gmazzap/b71e0d4eb99e0657d939bbc8d856bc7d that would be much appreciated.

**Comment by AndreKR** [comment]
Currently we're using path repositories when developing a dependency and TBH it works quite well (if you make sure not to run composer update when the wrong branch is checked out in the dependency).

However, recently our desire to commit `composer.lock` has grown stronger, in order to speed up builds and make them more reproducible.

Do I see this correctly that we cannot do that as long as this issue is not fixed because the local path repository will always end up in `composer.lock`?

**Comment by gmazzap** [comment]
@SanderSander If I understand correctly your plugin allow to link packages and link them (by creating a symlink) _after_ they are installed. So your plugin first let Composer pull the packages, and then deletes the code of each linked package and replaces it with the symlink.

If that is true, I don't think it is useful for my use case (and not only mine, according to comments in this issue).

Your plugin approach requires the package to exist in a repository that Composer understands, e.g. packagist. For me, the point of linked packages is to be able to develop them, and have them installed by Composer, _before_ the package is published, as I won't publish packages I've still to develop.

Moreover, IIIC, your plugin only provide "per project" linked packages, so if I have to link the same package in _$n_ different projects, I have to link them _$n_ times, because the link storage is per-project.

While this _might_ work for me, I think that the idea @Seldaek "approved" is to have "global" linked packages, so packages that linked packages are local to your system, and you can use them in all the projects you have in your system.

**Comment by AndreKR** [comment]
> For me, the point of linked packages is to be able to develop them, and have them installed by Composer, before the package is published, as I won't publish packages I've still to develop.

That is one use case but for me the time where a package is completely unpublished is relatively short so I could live with workarounds during that period.

My main use case is *after* the package has been published. At that time it is listed in composer.json and in composer.lock but I want to check it out manually and develop it locally, without making any changes to composer.json or composer.lock because those changes might get accidentally pushed to the repository.

**Comment by SanderSander** [comment]
> @SanderSander If I understand correctly your plugin allows to link packages and link them (by creating a symlink) _after_ they are installed. So your plugin first let Composer pull the packages, and then deletes the code of each linked package and replaces it with the symlink.
> 
> If that is true, I don't think it is useful for my use case (and not only mine, according to comments in this issue).
> 
> Your plugin approach requires the package to exist in a repository that Composer understands, e.g. packagist. For me, the point of linked packages is to be able to develop them, and have them installed by Composer, _before_ the package is published, as I won't publish packages I've still to develop.

At this point in time yes, the package needs to be installed. 
But it would be an easy change to link a package while it isn't installed, and I think it will be good to have this requirement. 
So I will add this feature in a few days :) 

> 
> Moreover, IIIC, your plugin only provide "per project" linked packages, so if I have to link the same package in _$n_ different projects, I have to link them _$n_ times, because the link storage is per-project.
> 
> While this _might_ work for me, I think that the idea @Seldaek "approved" is to have "global" linked packages, so linked packages would be local to your system, and you can use them in all the projects you have in your system without linking them again.

I think this should also be possible, but requires some more effort and testing. 
I always keep myself the ability to quickly unlink a package so I can see the original behavior en then link my development again back in. 

> > For me, the point of linked packages is to be able to develop them, and have them installed by Composer, before the package is published, as I won't publish packages I've still to develop.
> 
> That is one use case but for me the time where a package is completely unpublished is relatively short so I could live with workarounds during that period.
> 
> My main use case is _after_ the package has been published. At that time it is listed in composer.json and in composer.lock but I want to check it out manually and develop it locally, without making any changes to composer.json or composer.lock because those changes might get accidentally pushed to the repository.

This is exactly my use case also most of the time, and this is fully supported without altering the json or lock file.

**Comment by igortrinidad** [comment]
I'm not sure if it's exectly the same purpose of the issue, but I've always worried about a similar way to link local composer packages the same way as npm does with npm link, so I write this small node package that will scan your compose deps, let you decide which packages you are expecting to link locally and creates the linkage in your composer.json for the dependencies that you are working locally.

The purpose and usage it's very simple, I hope it helps somebody 

[https://github.com/igortrinidad/composer-link-local](https://github.com/igortrinidad/composer-link-local)
